File: Task/Decision-tables/Go/decision-tables.go
package main

import (
    "errors"
    "fmt"
    "os"
)

type dtText struct {
    rules, text string
}

var ptText = []dtText{
    {"YYYYNNNN", "Printer does not print"},
    {"YYNNYYNN", "A red light is flashing"},
    {"YNYNYNYN", "Printer is unrecognised"},
    {"--------", ""},
    {"  X     ", "Check the power cable"},
    {"X X     ", "Check the printer-computer cable"},
    {"X X X X ", "Ensure printer software is installed"},
    {"XX  XX  ", "Check/replace ink"},
    {" X X    ", "Check for paper jam"},
}

type dtMap map[string][]string

func compileDT(t []dtText) (dtMap, error) {
    if len(t) == 0 {
        return nil, errors.New("Empty decision table")
    }
    var conditions, actions []dtText
    ruleColumns := len(t[0].rules)
    for i, row := range t {
        if len(row.rules) != ruleColumns {
            return nil, errors.New("Inconsistent number of rule columns")
        }
        if len(row.text) == 0 {
            if conditions != nil {
                return nil, errors.New("Multple separator lines")
            }
            if i == 0 {
                return nil, errors.New("No conditions specified")
            }
            if i == len(t)-1 {
                return nil, errors.New("No actions specified")
            }
            conditions = t[:i]
            actions = t[i+1:]
        }
    }
    if conditions == nil {
        return nil, errors.New("Missing separator line")
    }
    m := make(map[string][]string, ruleColumns)
    kb := make([]byte, len(conditions))
    for col := 0; col < ruleColumns; col++ {
        for i, c := range conditions {
            kb[i] = c.rules[col]
        }
        key := string(kb)
        for _, a := range actions {
            if a.rules[col] != ' ' {
                m[key] = append(m[key], a.text)
            }
        }
    }
    return m, nil
}

func init() {
    var err error
    if ptMap, err = compileDT(ptText); err != nil {
        fmt.Println(err)
        os.Exit(1)
    }
}

var ptMap dtMap

func main() {
    for _, a := range ptMap["NYY"] {
        fmt.Println(a)
    }
}

File: Task/Decision-tables/R/decision-tables.r
library(e1071)

conditions <- `colnames<-`(bincombinations(3), c("prints", "redlight", "recognised"))
actions <- c("power"="Check the power cable",
             "usb"="Check the printer-computer cable",
             "software"="Ensure printer software is installed",
             "ink"="Check/replace ink",
             "jam"="Check for paper jam")

prompts <- c("Does the printer print? Y/N: ",
             "Is a red light flashing? Y/N: ",
             "Is the printer recognised by the computer? Y/N: ")

troubleshooter <- data.frame("power"=c(1,0,0,0,0,0,0,0),
                             "usb"=c(1,0,1,0,0,0,0,0),
                             "software"=c(1,0,1,0,1,0,1,0),
                             "ink"=c(0,0,1,1,0,0,0,1),
                             "jam"=c(0,1,0,1,0,0,0,0))

troubleshooter <- cbind(conditions, troubleshooter)

troubleshoot_printer <- function(){
  read_input <- function(p) readline(prompt=p) |> tolower()
  input2bool <- function(s) switch(s, "y"=1, "n"=0, stop("Input must be Y or N"))
  answers <- sapply(prompts, read_input)
  bools <- sapply(answers, input2bool)
  actions_list <- subset(troubleshooter, prints==bools[1]&redlight==bools[2]&recognised==bools[3])[4:8]
  actions_raw <- Filter(`+`, actions_list) |> names()
  cat("Take the following actions:\n")
  if(length(actions_raw)==0) cat("No actions required!\n")
  writeLines(actions[actions_raw])
}

troubleshoot_printer()
